src/py/flwr/server/state/sqlite_state.py

```diff
@@ -0,0 +1,582 @@
+    -- FOREIGN KEY(task_id) REFERENCES node(id)
+);
+"""
+
+
+SQL_CREATE_TABLE_TASK_RES = """
+CREATE TABLE IF NOT EXISTS task_res(
+    task_id                 TEXT UNIQUE,
+    group_id                TEXT,
+    workload_id             TEXT,
+    producer_anonymous      BOOLEAN,
+    producer_node_id        INTEGER,
+    consumer_anonymous      BOOLEAN,
+    consumer_node_id        INTEGER,
+    created_at              TEXT,
+    delivered_at            TEXT,
+    ttl                     TEXT,
+    ancestry                TEXT,
+    legacy_server_message   BLOB,
+    legacy_client_message   BLOB
+    -- FOREIGN KEY(task_id) REFERENCES node(id)
+);
+"""
+
+DictOrTuple = Union[Tuple[Any], Dict[str, Any]]
+
+
+class SqliteState(State):
+    """SQLite based state implemenation."""
+
+    def __init__(
+        self,
+        database_path: str,
+    ) -> None:
+        """Initialize an SqliteState.
+
+        Parameters
+        ----------
+        database : (path-like object)
+            The path to the database file to be opened. Pass ":memory:" to open
+            a connectionto a database that is in RAM instead of on disk.
+        """
+        self.database_path = database_path
+        self.conn: Optional[sqlite3.Connection] = None
+
+    def initialize(self) -> List[Tuple[str]]:
+        """Create tables if they don't exist yet."""
+        self.conn = sqlite3.connect(self.database_path)
+        self.conn.row_factory = dict_factory
+        self.conn.set_trace_callback(lambda query: log(1, query))
+        cur = self.conn.cursor()
+
+        # Create each table if not exists queries
+        cur.execute(SQL_CREATE_TABLE_TASK_INS)
+        cur.execute(SQL_CREATE_TABLE_TASK_RES)
+        cur.execute(SQL_CREATE_TABLE_NODE)
+
+        res = cur.execute("SELECT name FROM sqlite_schema;")
+
+        return res.fetchall()
+
+    def query(
+        self,
+        query: str,
+        data: Optional[Union[List[DictOrTuple], DictOrTuple]] = None,
+    ) -> List[Dict[str, Any]]:
+        """Run SQLQuery."""
+        if self.conn is None:
+            raise Exception("State is not initialized.")
+
+        if data is None:
+            data = []
+
+        # Cleanup whitespace to make the logs nicer
+        query = re.sub(r"\s+", " ", query)
+
+        try:
+            with self.conn:
+                if (
+                    len(data) > 0
+                    # pylint: disable-next=C0123
+                    and (type(data) == tuple or type(data) == list)
+                    # pylint: disable-next=C0123
+                    and (type(data[0]) == tuple or type(data[0]) == dict)
+                ):
+                    rows = self.conn.executemany(query, data)
+                else:
+                    rows = self.conn.execute(query, data)
+
+                # Extract results before commiting to support
+                #   INSERT/UPDATE ... RETURNING
+                # style queries
+                result = rows.fetchall()
```

Comment:
The comment explains why this happens before the commit, but where does the actual commit happen?

Reply:
So this was before I realized the connection could be used as a context manager, although it still makes sense. Basically, what's happing when `with self.conn:` is used is that a "BEGIN" query is executed, and at the exit of the context manager a "COMMIT" is executed. SQLite requires the "COMMIT" to write any change to disk.

Reply:
Got it, thanks!